### ISSUE TITLE
Creating new service on swarm doesn't publish ports on Centos7

### ISSUE THREAD

**Issue by virtuman** [docker/docker#26235]
**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 6
Server Version: 1.12.1
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null overlay host
Swarm: active
 NodeID: 4aw23p0m4s1hazwr3jw3h5lr1
 Is Manager: true
 ClusterID: dowu6j9g030etwhzxd1rzv5b7
 Managers: 4
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.16.130.111
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.7.0-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 62.81 GiB
Name: www5.strippeddomain.com
ID: SSGC:WWQK:QTW2:T3OC:GAQW:RCZ7:YXAY:I5HY:XDWK:XIOS:BZWE:RTG6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical machines with dual network cards. one for external, and one for internal (cross-server communication) networks

**Steps to reproduce the issue:**
these steps are taken directly from Docker Online webinar 42
1. Clean install *CentOS7* or any version up to current *CentOS Linux release 7.2.1511 (Core)*
2. install docker engine 1.12.*
3. docker network create -d overlay collabnet
4. docker service create --name wordpressdb1 --network collabnet -e MYSQL_ROOT_PASSWORD=mysql123 -e MYSQL_DATABASE=wordpress --replicas 2 mysql:latest
5. docker service create -e WORDPRESS_DB_HOST=wordpressdb1 -e WORDPRESS_DB_PASSWORD=mysql123 --network collabnet --replicas 3 --name wordpressapp --publish 81:80/tcp wordpress:latest

**Describe the results you received:**
Containers for all services with all replicas get spun up correctly, but port 81 does not become accessible

**Describe the results you expected:**
Expected to be able to access port :81 on any of the nodes within the swarm cluster to be able to access wordpress interface, but none of the nodes had this port exposed.

**Additional information you deem important (e.g. issue happens only occasionally):**
I suspect some of the things, not sure if any of them make sense:
1. it looks like ipv6 is enabled, and possibly somehow interfering with ipv4 port assignment
2. it's possible that iptables on centos is the culprit 
3. i basically cannot start any of the containers that would have port published, none of them will publish the port
4. while docker service won't publish ports, regular docker run continues to publish ports for newly spun up services correctly
5. according to reply via email - @mgoelzer confirms that this is a bug:

> If I'm understanding correct, a service started like `docker service create -p 12345:12345 whatever-image /some/command/that/listens/on/12345` does not expose :12345 on all 8 of your nodes, right?  Yes, that is definitely a bug.  Can you open an issue about it on docker/docker and @-mention me?  (@mgoelzer)

Possibly the issue has to do with iptables because of this:
```
systemctl status firewalld
● firewalld.service - firewalld - dynamic firewall daemon
   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; enabled; vendor preset: enabled)
   Active: active (running) since Thu 2016-09-01 01:36:25 UTC; 5h 34min ago
 Main PID: 18703 (firewalld)
   Memory: 32.7M
   CGroup: /system.slice/firewalld.service
           └─18703 /usr/bin/python -Es /usr/sbin/firewalld --nofork --nopid

Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C FORWARD -o docker0 -j DOCKER' failed: iptables: No chain/target/match by that name.
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C FORWARD -j DOCKER-ISOLATION' failed: iptables: No chain/target/match by that name.
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C POSTROUTING -s 172.19.0.0/16 ! -o docker_gwbridge -j MASQUERADE' failed: iptables: No chain/target/match by that name.
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C DOCKER -i docker_gwbridge -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -D FORWARD -i docker_gwbridge -o docker_gwbridge -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C FORWARD -i docker_gwbridge -o docker_gwbridge -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C FORWARD -i docker_gwbridge ! -o docker_gwbridge -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C FORWARD -o docker_gwbridge -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C FORWARD -o docker_gwbridge -j DOCKER' failed: iptables: No chain/target/match by that name.
Sep 01 01:36:26 www5.domain123.com firewalld[18703]: 2016-09-01 01:36:26 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -D FORWARD -i docker0 -o docker0 -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).
```

**Comment by mrjana** [comment]
@virtuman can you provide a dump of `iptables -t nat -nvL` and `iptables -nvL` from the host where you are having trouble? Also can you post any errors you see from dockerd logs? From your logs it looks your firewalld is complaining about a lot of basic iptable rules. Is it possible for you retry the whole thing with firewalld disabled?

**Comment by virtuman** [comment]
of course, here you go.. some IPs are changed but mainly everything is here:

```
[root@www5 ~]# iptables -t nat -nvL
Chain PREROUTING (policy ACCEPT 1834K packets, 102M bytes)
 pkts bytes target     prot opt in     out     source               destination
1834K  102M PREROUTING_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1834K  102M PREROUTING_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1834K  102M PREROUTING_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1816K  101M DOCKER     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT 1800K packets, 100M bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy ACCEPT 1621K packets, 97M bytes)
 pkts bytes target     prot opt in     out     source               destination
1621K   97M OUTPUT_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
 185K   11M DOCKER     all  --  *      *       0.0.0.0/0           !127.0.0.0/8          ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT 1632K packets, 98M bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 MASQUERADE  all  --  *      !docker_gwbridge  172.19.0.0/16        0.0.0.0/0
 6496  405K MASQUERADE  all  --  *      !docker0  172.17.0.0/16        0.0.0.0/0
1632K   98M POSTROUTING_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1632K   98M POSTROUTING_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1632K   98M POSTROUTING_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 MASQUERADE  tcp  --  *      *       172.17.0.5           172.17.0.5           tcp dpt:80

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     all  --  docker_gwbridge *       0.0.0.0/0            0.0.0.0/0
    0     0 RETURN     all  --  docker0 *       0.0.0.0/0            0.0.0.0/0
    5   276 DNAT       tcp  --  !docker0 *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8000 to:172.17.0.5:80

Chain OUTPUT_direct (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain POSTROUTING_ZONES (1 references)
 pkts bytes target     prot opt in     out     source               destination
1632K   98M POST_drop  all  --  *      +       0.0.0.0/0            0.0.0.0/0

Chain POSTROUTING_ZONES_SOURCE (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain POSTROUTING_direct (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain POST_drop (1 references)
 pkts bytes target     prot opt in     out     source               destination
1632K   98M POST_drop_log  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1632K   98M POST_drop_deny  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1632K   98M POST_drop_allow  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain POST_drop_allow (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain POST_drop_deny (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain POST_drop_log (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain PREROUTING_ZONES (1 references)
 pkts bytes target     prot opt in     out     source               destination
1834K  102M PRE_drop   all  --  +      *       0.0.0.0/0            0.0.0.0/0

Chain PREROUTING_ZONES_SOURCE (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain PREROUTING_direct (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain PRE_drop (1 references)
 pkts bytes target     prot opt in     out     source               destination
1834K  102M PRE_drop_log  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1834K  102M PRE_drop_deny  all  --  *      *       0.0.0.0/0            0.0.0.0/0
1834K  102M PRE_drop_allow  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain PRE_drop_allow (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain PRE_drop_deny (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain PRE_drop_log (1 references)
 pkts bytes target     prot opt in     out     source               destination
[root@www5 ~]#
```












```
[root@www5 ~]# iptables -nvL
Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
  61M   44G ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
 197K   12M ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
2005K  162M INPUT_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
 114K   58M INPUT_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
 114K   58M INPUT_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0
    4   232 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
 677K 1429M DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  docker_gwbridge !docker_gwbridge  0.0.0.0/0            0.0.0.0/0
 536K 1420M DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0
 525K 1419M ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
 141K 9349K ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
11390  683K ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
    3   209 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
    0     0 FORWARD_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FORWARD_IN_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FORWARD_IN_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FORWARD_OUT_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FORWARD_OUT_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited
    0     0 DROP       all  --  docker_gwbridge docker_gwbridge  0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 52M packets, 105G bytes)
 pkts bytes target     prot opt in     out     source               destination
  52M  105G OUTPUT_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination
   23  1334 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.5           tcp dpt:80

Chain DOCKER-ISOLATION (1 references)
 pkts bytes target     prot opt in     out     source               destination
 677K 1429M RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain FORWARD_IN_ZONES (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 FWDI_drop  all  --  +      *       0.0.0.0/0            0.0.0.0/0

Chain FORWARD_IN_ZONES_SOURCE (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD_OUT_ZONES (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 FWDO_drop  all  --  *      +       0.0.0.0/0            0.0.0.0/0

Chain FORWARD_OUT_ZONES_SOURCE (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD_direct (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FWDI_drop (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 FWDI_drop_log  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FWDI_drop_deny  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FWDI_drop_allow  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain FWDI_drop_allow (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FWDI_drop_deny (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FWDI_drop_log (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FWDO_drop (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 FWDO_drop_log  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FWDO_drop_deny  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 FWDO_drop_allow  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain FWDO_drop_allow (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FWDO_drop_deny (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FWDO_drop_log (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain INPUT_ZONES (1 references)
 pkts bytes target     prot opt in     out     source               destination
 114K   58M IN_drop    all  --  +      *       0.0.0.0/0            0.0.0.0/0

Chain INPUT_ZONES_SOURCE (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain INPUT_direct (1 references)
 pkts bytes target     prot opt in     out     source               destination
 850K   44M ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp multiport dports 80,443,28777 /* http/https/logio */
    8   420 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp multiport dports 2377 /* docker swarm port */
    0     0 ACCEPT     udp  --  *      *       0.0.0.0/0            0.0.0.0/0            udp multiport dports 7946,4789 /* docker swarm port */
1039K   59M ACCEPT     tcp  --  eth1   *       0.0.0.0/0            0.0.0.0/0            tcp /* open all ports on local network interface */
    0     0 ACCEPT     tcp  --  lo     *       0.0.0.0/0            0.0.0.0/0            tcp /* lo accept */
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED /* established */
    0     0 ACCEPT     tcp  --  *      *       172.16.0.0/12        0.0.0.0/0            tcp /* allow local network access */
    0     0 ACCEPT     udp  --  *      *       172.16.0.0/12        0.0.0.0/0            /* allow local network access */
    3   160 ACCEPT     tcp  --  *      *       10.10.24.27       0.0.0.0/0            tcp /* office 2 */
    0     0 ACCEPT     udp  --  *      *       10.10.24.27       0.0.0.0/0            /* office 2 */
    0     0 ACCEPT     tcp  --  *      *       13.89.16.34        0.0.0.0/0            tcp /* office 3 */
    0     0 ACCEPT     udp  --  *      *       13.89.16.34        0.0.0.0/0            /* office 3 */
    0     0 ACCEPT     tcp  --  *      *       7.10.23.30       0.0.0.0/0            tcp /* office */
    0     0 ACCEPT     udp  --  *      *       7.10.23.30       0.0.0.0/0            /* office */
    0     0 ACCEPT     tcp  --  *      *       2.20.2.4/29     0.0.0.0/0            tcp multiport dports 161 /*  */
    0     0 ACCEPT     udp  --  *      *       2.20.2.4/29     0.0.0.0/0            multiport dports 161 /*  */
    0     0 ACCEPT     tcp  --  *      *       14.17.13.4/29    0.0.0.0/0            tcp multiport dports 161 /* */
    0     0 ACCEPT     udp  --  *      *       14.17.13.4/29    0.0.0.0/0            multiport dports 161 /*  */

Chain IN_drop (1 references)
 pkts bytes target     prot opt in     out     source               destination
 114K   58M IN_drop_log  all  --  *      *       0.0.0.0/0            0.0.0.0/0
 114K   58M IN_drop_deny  all  --  *      *       0.0.0.0/0            0.0.0.0/0
 114K   58M IN_drop_allow  all  --  *      *       0.0.0.0/0            0.0.0.0/0
 114K   58M DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain IN_drop_allow (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain IN_drop_deny (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain IN_drop_log (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT_direct (1 references)
 pkts bytes target     prot opt in     out     source               destination
[root@www5 ~]#
```

**Comment by nickvanw** [comment]
I believe I'm seeing the same bug on my end as well:

**Output of `docker version`:**

```
root@swarm-node03:~# docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```


**Output of `docker info`:**

```
root@swarm-node03:~# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-253:1-131847-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 118.1 MB
 Data Space Total: 107.4 GB
 Data Space Available: 40.23 GB
 Metadata Space Used: 782.3 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.110 (2015-10-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay null host bridge
Swarm: active
 NodeID: 9cvuvmpi9jc1dxo0imxmfc52r
 Is Manager: true
 ClusterID: 6vc8gbla0fnwyv5e2imd80md2
 Managers: 1
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: ip.address.here
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-34-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.953 GiB
Name: swarm-node03.something.goes.here
ID: GVUQ:PRSY:5C2Y:6XQZ:6FFZ:IGDK:FWKP:BSBH:AFWW:WLBB:RJME:GNYX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

These are DigitalOcean Droplets running 16.04 set up following the guide, without any current running containers. 

**Steps to reproduce the issue:**
1. Create 3 DigitalOcean Droplets running Ubuntu 16.04.1 w/ Private Networking
2. Install the latest Docker 1.12.1
3. Join them into a swarm cluster, verify it's working:
```
root@swarm-node03:~# docker node ls
ID                           HOSTNAME                           STATUS  AVAILABILITY  MANAGER STATUS
0emcew1av540uxifaupg92rq6    swarm-node02.something.here  Ready   Active
9cvuvmpi9jc1dxo0imxmfc52r *  swarm-node03.something.here  Ready   Active        Leader
deouq7l2xp5a3xa6qqmzls9ic    swarm-node01.something.here  Ready   Active
```
4. Create a global Docker Service that exposes port 80:
```
root@swarm-node03:~# docker service create --network=overlay-net --mode=global -p 80:80 tutum/hello-world
82q1kfnhdwxior2juslgbr0pe
```
5. Attempt to reach port 80 on all of the Swarm nodes in the cluster:
```
root@swarm-node03:~# curl localhost
```
On each node



**Describe the results you received:**
doing a `curl localhost` will work on certain nodes, and will fail on other nodes, which are not able to reach the container running on them:

Working:
```
root@swarm-node03:~# curl localhost
<html>
<head>
       	<title>Hello world!</title>
       	<link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
       	<style>
       	body {
       		background-color: white;
       		text-align: center;
       		padding: 50px;
       		font-family: "Open Sans","Helvetica Neue",Helvetica,Arial,sans-serif;
       	}

       	#logo {
       		margin-bottom: 40px;
       	}
       	</style>
</head>
<body>
       	<img id="logo" src="logo.png" />
       	<h1>Hello world!</h1>
       	<h3>My hostname is ba7eeba139bd</h3>   	</body>
</html>
```

Not Working:

```
root@swarm-node01:~# docker ps
CONTAINER ID        IMAGE                      COMMAND                  CREATED             STATUS              PORTS               NAMES
ae3329e58250        tutum/hello-world:latest   "/bin/sh -c 'php-fpm "   5 minutes ago       Up 5 minutes        80/tcp              tender_bell.0.a995xjh5zx9ol5cot90u21qdw
root@swarm-node01:~# curl localhost
curl: (7) Failed to connect to localhost port 80: No route to host
```


**Describe the results you expected:**
It's expected that every node in the Swarm cluster will respond on localhost port 80 and balance across all of the containers in the Docker service. 

**Additional information you deem important (e.g. issue happens only occasionally):**
I haven't been able to pin down if there is a pattern to which node(s) are not serving traffic properly - as my cluster is only 3 nodes, it's difficult to see if the behavior is deterministic. 


I'm happy to paste the output of `iptables` or anything else - I'll leave the Droplets around as they are. I'm also happy to give anyone access to the Droplets if any live debugging will be helpful

**Comment by virtuman** [comment]
Can you confirm that you are also unable to reach any of the hosts from the external connections on port 80 in the described scenario, ie:

`curl swarm-node03.something.here`

for me the result is: port is closed on all nodes

**Comment by liuxiaopiao** [comment]
met the same problem

**Comment by virtuman** [comment]
Not sure if this is related, all of my servers have multiple network adaptors:
eth0 for public network
eth1 for local / internal network connectivity

swarm is started to listen and connect on internal network interface, however curl'ing from another server on a local/internal network - it is still unable to connect. localhost will sometimes be able to access service on the correct port, but accessing from non-localhost wouldn't work at all.

I tried  the following steps to no avail:
1) service firewalld stop
2) service iptables start   (with NO RULES)
3) ip link delete docker0
4) service docker restart
DID NOT WORK.

Then I tried:
1) service firewalld stop
2) service iptables stop
3) ip link delete docker0
4) service docker restart

everything worked as expected, but I had no firewall running. So I tried to delete docker0 interface again and restarting DOCKER with iptables or firewalld and in neither scenario it worked correctly.

**Comment by drsolaris** [comment]
I think I have the same issue. Currently, I am only using a single machine and have two networks with containers attached. The first is a traditional bridge network (172.18.0.0/16), with containers attached and publishing ports; this works fine. The second is an overlay network (172.19.0.0/16) for use with swarm. Unfortunately, when creating a service within the overlay network and publishing a port (8000), the port is not correctly routed from the host into the service. Curling 172.19.0.2:8000 from the host works, but curling 192.168.1.4:8000 (the host address) from other machines in the external network doesn't.

Forgive me if the following is nonsense, I know very little about iptables, but the only major difference I can find is that there appears to be 3 sets of iptables rules regarding the bridge network IP range, and only 2 for the overlay network.

First, the bridge network:

```sh
> sudo iptables -t nat -L -n | grep -E '18\.'
MASQUERADE  all  --  172.18.0.0/16        0.0.0.0/0
MASQUERADE  tcp  --  172.18.0.3           172.18.0.3           tcp dpt:8000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:2181 to:172.18.0.3:8000
```

The first is under 'Chain PREROUTING', second 'Chain POSTROUTING', and third 'Chain DOCKER'. Now the overlay network:

```sh
> sudo iptables -t nat -L -n | grep -E '19\.'
MASQUERADE  all  --  172.19.0.0/16        0.0.0.0/0
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8000 to:172.19.0.2:8000
```

The first is under 'Chain PREROUTING' and the second 'Chain DOCKER-INGRESS'.

Am I right in thinking that a required MASQUERADE POSTROUTING record for the overlay network hasn't been created?

**Comment by toughIQ** [comment]
I can confirm the same behavior when using RHEL7.2 and 1.12.1
`docker run -p` works as expected. 
`docker service create --publish` fails. `netstat` shows a port binding, but requests fail.

**Comment by sanimej** [comment]
@virtuman I created a new 3 node Cent OS cluster in AWS and started a service with 3 replicas. It works as expected; accessing the published port rotates through all the 3 backend instances.

```
[root@ip-172-31-41-169 ~]# docker service create --network ov1 --name web --replicas 3 -p 80:8000 jwilder/whoami
9dccvho830puz6h2dk8yok2ck

node 1:

[root@ip-172-31-41-169 ~]# curl 127.0.0.01
I'm 72fd34b867e0
[root@ip-172-31-41-169 ~]# curl 127.0.0.01
I'm 625585c13d57
[root@ip-172-31-41-169 ~]# curl 127.0.0.01
I'm df4fb12def55
[root@ip-172-31-41-169 ~]# 

node 2:

[root@ip-172-31-39-96 ~]# curl 127.0.0.1 
I'm 72fd34b867e0
[root@ip-172-31-39-96 ~]# curl 127.0.0.1 
I'm df4fb12def55
[root@ip-172-31-39-96 ~]# curl 127.0.0.1 
I'm 625585c13d57
[root@ip-172-31-39-96 ~]# 

node 3:

[root@ip-172-31-32-62 ~]# curl 127.0.0.1
I'm df4fb12def55
[root@ip-172-31-32-62 ~]# curl 127.0.0.1
I'm 625585c13d57
[root@ip-172-31-32-62 ~]# curl 127.0.0.1
I'm 72fd34b867e0
```
I don't have firewalld running. But looks like you are seeing the issue with & without firewalld.

This is stock Cent OS install which has kernel 3.10.0-327.10.1.el7.x86_64. You have a newer kernel version, 4.7.0-1.el7.elrepo.x86_64. Have you tried on a Cent OS with 3.10.x kernel ? 

Lets use this issue to track this and close the earlier issue you have  opened for the same symptom, #24882 ?

**Comment by virtuman** [comment]
Curl'ing from localhost works just fine, but not a real environment usecase. The port remains closed from public access.

Just as described in the original issue and the replies in this issue:
Accessing port through loopback (localhost or 127.0.0.1) seems to work fine, however - cannot access port from the external connection (or from any other server), as in your example, i'd be trying this:
`curl 172.31.41.169`

> from node [root@ip-172-31-39-96 ~]# curl 127.0.0.1

**Comment by sanimej** [comment]
curl from a different node works as well..
```
[centos@ip-172-31-39-96 ~]$ curl 172.31.41.169
I'm df4fb12def55
[centos@ip-172-31-39-96 ~]$ curl 172.31.41.169
I'm 72fd34b867e0
[centos@ip-172-31-39-96 ~]$ curl 172.31.41.169
I'm 625585c13d57
[centos@ip-172-31-39-96 ~]$ 
```
I will try with firewalld enabled to see if it makes a difference.

**Comment by virtuman** [comment]
can you post `sysctl -a` output ? or what other relevant things could we be looking into checking?

**Comment by virtuman** [comment]
**docker node ls**
```
ID                           HOSTNAME                 STATUS  AVAILABILITY  MANAGER STATUS
4aw23p0m4s1hazwr3jw3h5lr1 *  www1.com  Ready   Active        Reachable
80y9oxv5tdgcwy1vnnb20x2pf    www2.com  Ready   Active        Leader
bejmawzxaa692mq3gqgkr81d5    www3.com  Ready   Active
e2cddzipbwbaq3k5tmz7gpnre    www4.com  Down    Active        Unreachable
```

**start up services**
```
docker service create --name wordpressdb1 --network collabnet -e MYSQL_ROOT_PASSWORD=mysql123 -e MYSQL_DATABASE=wordpress --replicas 2 mysql:latest
docker service create -e WORDPRESS_DB_HOST=wordpressdb1 -e WORDPRESS_DB_PASSWORD=mysql123 --network collabnet --replicas 3 --name wordpressapp --publish 85:80 wordpress:latest
```

**docker ps**
```
CONTAINER ID        IMAGE                              COMMAND                  CREATED             STATUS              PORTS                           NAMES
7078bdc696f9        wordpress:latest                   "/entrypoint.sh apach"   About an hour ago   Up About an hour    80/tcp                          wordpressapp.3.96yxfjftofvfldeykmjgb8ree
```

**docker service ls**
```
ID            NAME          REPLICAS  IMAGE             COMMAND
4uf5g3ouwzw7  wordpressdb1  2/2       mysql:latest
av2g7rp0h726  wordpressapp  3/3       wordpress:latest
```

**on loopback - works properly**
lynx 127.0.0.1:85

**from 172.16.130.116 - works properly**
lynx 172.16.130.116:85

**With FIREWALLD running, but accept all traffic from local LAN interface: **
**from another computer on same LAN  - FAIL**
lynx 172.16.130.116:85

and it will fail from any other computer whether on LAN or WAN, on primary or secondary network interface as well from public internet with firewalld whitelisting the request port.

Same failures with firewalld disabled.
kernel versions attempted: 
stock centos7 3.10.* kernel
multiple upgrade to current 3.10.*
multiple 4.7.* 
all fail

**Comment by virtuman** [comment]
Just noticed a couple of warning on `docker info` that i must have overlooked before:

```
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

so i modified using systcl -A net.bridge......=1
stopped firewalld and restarted docker, i am now able to connect with firewalld correctly and everything works, but as soon as i start firewalld back up - port is no longer published.

**Comment by virtuman** [comment]
after starting up firewalld i checked the default input/output/forward polcies, they are all set to ACCEPT:

`(iptables -L INPUT; iptables -L OUTPUT; iptables -L FORWARD) | grep policy`
> Chain INPUT (policy ACCEPT)
> Chain OUTPUT (policy ACCEPT)
> Chain FORWARD (policy ACCEPT)

**Comment by virtuman** [comment]
when you test with firewalld enabled, could you also paste output of:
`iptables -L FORWARD -nv`

thank you

**Comment by perigee** [comment]
Met same issue. Launching service nginx (only one) on swarm, the port is only publish on the node service running, then update the service nginx, the nginx service will running on another node, while the service can only be reached on previous host ip.

**Comment by sanimej** [comment]
@virtuman In my cluster I enabled firewalld on one of the worker nodes. I wasn't able to access the published port of the task only on that node. I just brought up firewalld on this node without any change to the zone/interface membership configs. It looks like the inter-host vxlan encapsulated packets are getting dropped by this rule in the INPUT chain..

`REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited`

For quick & dirty debugging if this rule is removed, accessing the task on that node works ok. But what you should do is use the firewalld commands to make sure the following ports are not blocked, 7946/udp & tcp (for network control plane) and 4789/udp (for vxlan data). 

Another option would be to disable firewalld if thats acceptable. After disabling firewalld restart the docker daemon because disabling firewall nukes all the iptables rule docker has setup before.

**Comment by virtuman** [comment]
For some reason neither of the two options seem to work in my case:

**Here's what I tried so far:**
1) stop firewalld on all nodes
2) restart docker on all nodes
3) start up firewalld on only one node
4) just in case - restart docker again on this node

**Result - Fail:**
1) can access poublished port on all nodes except the node that has firewalld running

Find the rule:
**# iptables -L -v -n --line-numbers**
```
Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination
1     1163 1224K ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
2        4   240 ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
3       49  3232 INPUT_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
4        7   920 INPUT_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
5        7   920 INPUT_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
6        0     0 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0
7        4   224 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited
```
Delete the rule in INPUT chain
**iptables -D INPUT 7**

Then verify that the rule is deleted:
**# iptables -L -v -n --line-numbers**
```
Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination
1    19943   19M ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
2       73  4380 ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
3      932 75440 INPUT_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
4       51 26264 INPUT_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
5       51 26264 INPUT_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
6        0     0 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0
```

That doesn't make it work.


Then I tried second approach, i went very "verbose" on unblocking ports (just in case)
```

# TCP & UDP unblock port 4789 & port 7946 on ipv4
firewall-cmd --direct --add-rule ipv4 filter INPUT 5 -p tcp -m tcp -m multiport --dports 4789  -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv4 filter INPUT 5 -p tcp -m tcp -m multiport --dports 4789 -j ACCEPT
firewall-cmd --direct --add-rule ipv4 filter INPUT 5 -p udp -m multiport --dports 4789 -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv4 filter INPUT 5 -p udp -m multiport --dports 4789  -j ACCEPT

firewall-cmd --direct --add-rule ipv4 filter INPUT 5 -p tcp -m tcp -m multiport --dports 7946  -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv4 filter INPUT 5 -p tcp -m tcp -m multiport --dports 7946 -j ACCEPT
firewall-cmd --direct --add-rule ipv4 filter INPUT 5 -p udp -m multiport --dports 7946 -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv4 filter INPUT 5 -p udp -m multiport --dports 7946  -j ACCEPT


# TCP & UDP unblock port 4789 & port 7946 on ipv6
firewall-cmd --direct --add-rule ipv6 filter INPUT 5 -p tcp -m tcp -m multiport --dports 4789  -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv6 filter INPUT 5 -p tcp -m tcp -m multiport --dports 4789 -j ACCEPT
firewall-cmd --direct --add-rule ipv6 filter INPUT 5 -p udp -m multiport --dports 4789 -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv6 filter INPUT 5 -p udp -m multiport --dports 4789  -j ACCEPT

firewall-cmd --direct --add-rule ipv6 filter INPUT 5 -p tcp -m tcp -m multiport --dports 7946  -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv6 filter INPUT 5 -p tcp -m tcp -m multiport --dports 7946 -j ACCEPT
firewall-cmd --direct --add-rule ipv6 filter INPUT 5 -p udp -m multiport --dports 7946 -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv6 filter INPUT 5 -p udp -m multiport --dports 7946  -j ACCEPT
```


And still no go. I made sure to restart docker on all nodes (again, just in case). 

Any other suggestions ? Thank you.

**Comment by sanimej** [comment]
@virtuman When I had firewalld running on one of the nodes, I was probably accessing the service's published port through the IP of other nodes in the cluster. Task on the node with firewalld running was accessible after I removed the INPUT chain rule that I mentioned earlier.

But when I accessed the port with the IP of the node that has firewalld running it didn't work. This is probably what you are hitting too. I was using curl and getting the error `No route to host` which typically means an ICMP unreachable is being received. firewalld was adding a REJECT rule in the FORWARD chain as well.

`REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited`

After removing rule, accessing the published port using the IP of the node which has firewalld running works as well. Please give it a try.

Note that if you want to deploy swarm with firewalld I would recommend against directly tweaking the iptables rules and instead find a way to achieve the same using `firewall-cmd`

**Comment by virtuman** [comment]
Tried deleting FORWARD's reject rule `iptables -D INPUT 17`.

For some reason - for me it still didn't work. Any other ideas?

You

**Comment by sanimej** [comment]
`iptables -D INPUT 17` if this is the command you were executing its the wrong chain. The REJECT rule to be remove is in `FORWARD` chain.

**Comment by virtuman** [comment]
Sorry for confusion, it was a continuation of the previous commands that I showed. What I was trying to say is that I deleted _REJECT_ rule from both **INPUT** and **FORWARD** rules .

Any other thoughts?

**Comment by slawekgh** [comment]
I made several tests , 3 nodes swarm , centos 7.2 , docker 1.12.1-rc2  

with firewalld enabled - load balancing is not stable (sometimes works, sometimes not, no access from outside world) 
with firewalld disabled - the same , works a little bit better but still not stable 
with firewalld disabled and then docker service restarted - everything is ok 

here is full log:

```
3 x (CentOS Linux release 7.2.1511 (Core)) , docker version 1.12.1-rc2
```
My swarm cluster:
```
[root@cent03 ~]# docker node ls
ID                           HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS
1tcaanthfe6qwosacmwv9qqj8    cent05    Ready   Active
8sh4wc5p6b9lcbpatgz0urcq6 *  cent03    Ready   Active        Leader
ap5agizea1vn6bek364i33tbc    cent04    Ready   Active
```
swarm service created via:
```
docker service create --name web04 --replicas 2 --publish 8084:80 test_apache
```
tasks running on 2 nodes (cent04 & cent05):
```
[root@cent03 ~]# docker service  ps web04
ID                         NAME     IMAGE        NODE    DESIRED STATE  CURRENT STATE           ERROR
0z7o31xglaez9ghtyclaw1mn8  web04.1  test_apache  cent04  Running        Running 29 minutes ago
921o4axrpgivanhjjsa81trrs  web04.2  test_apache  cent05  Running        Running 29 minutes ago
```
Container names are 4f68a73f3332 and 3543978e30ea

From cent03 I can curl 127.0.0.1:8084  but it returns only one container - seems no LoadBalancing works:
```
[root@cent03 ~]# curl 127.0.0.1:8084
4f68a73f3332
```
From cent04 curl returns also the same container
```
[root@cent04 ~]# curl 127.0.0.1:8084
4f68a73f3332
```
BUT from cent05 curl returns 2nd container 
```
[root@cent05 ~]# curl 127.0.0.1:8084
3543978e30ea
```

No way to curl from external machine and on nodes from each other 


**When I disable firewalld (via systemctl stop firewalld & systemctl disable firewalld)**

a) all of the local curl (127.0.0.1) stopped working
b) after creation of new service 
```
[root@cent03 ~]# docker service create --name web05 --replicas 2 --publish 8085:80 test_apache
```

Swarm put 2 containers on machines:
```
[root@cent03 ~]# docker service  ps web05
ID                         NAME     IMAGE        NODE    DESIRED STATE  CURRENT STATE               ERROR
aob48mcy74vhyn7nr1o19bd70  web05.1  test_apache  cent03  Running        Running about a minute ago
e8h73ornao5q7p1w6l9ewqckq  web05.2  test_apache  cent03  Running        Running about a minute ago
```

from cent03 seems that curl working and there is also load-balancing:
```
[root@cent03 ~]# curl 127.0.0.1:8085
c78a5521865e
[root@cent03 ~]# curl 127.0.0.1:8085
55e7925c2ea0
```

from cent04 - the same
```
[root@cent04 ~]# curl 127.0.0.1:8085
c78a5521865e
[root@cent04 ~]# curl 127.0.0.1:8085
55e7925c2ea0
```

BUT NOT FROM cent05:

```
[root@cent05 ~]# curl 127.0.0.1:8085
curl: (7) Failed connect to 127.0.0.1:8085; Connection refused
```
From outside world curl works - but only to cent03 and cent04 , to cent05 doesnt work 



**as a next step i reload all docker daemons - everything works now ok**

**Comment by mendeni** [comment]
We are continuing to see this issue with 1.12.2-rc1 on Debian jessie with a 3.12.21 kernel.
```
# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.2-rc1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: active
 NodeID: bms42xsdsbv3lahs1i5ynpr8k
 Is Manager: true
 ClusterID: 5ctkuppqmls0okmypggfbni3s
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.1.251
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.18.21
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 256 MiB
Name: test01
ID: MMHP:NEYA:2UFU:PQPQ:QIRR:VRD4:3TLW:N6GX:76ZE:GESH:OKYM:LZ4U
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8

# docker swarm init
Swarm initialized: current node (c4qb0ch0n52nd7hxkn8zg49ez) is now a manager.
...

# docker service create --name my_web --replicas 3 --publish 8080:80 nginx
bee8dzp9tgwvoulw7l28y3cna

# docker service ls
ID            NAME    REPLICAS  IMAGE  COMMAND
9b4o5ngf7xx1  my_web  0/3       nginx
```
A broken service will have these errors logged:
```
Sep 27 20:58:30 test01 dockerd[99]: time="2016-09-27T20:58:30.659574742Z" level=error msg="Failed to add ingress: failed to find gateway bridge interface name for <nil>: numerical result out of range"
Sep 27 20:58:30 test01 dockerd[99]: time="2016-09-27T20:58:30.659812688Z" level=error msg="Failed to create real server 10.255.0.7 for vip 10.255.0.4 fwmark 2865 in sb ingress-sbox: no such process"
Sep 27 20:58:30 test01 dockerd[99]: time="2016-09-27T20:58:30.691847913Z" level=warning msg="Could not rollback container connection to network ingress"
Sep 27 20:58:30 test01 dockerd[99]: time="2016-09-27T20:58:30.706289781Z" level=info msg="Failed to delete real server 10.255.0.7 for vip 10.255.0.4 fwmark 2865: no such process"
Sep 27 20:58:30 test01 dockerd[99]: time="2016-09-27T20:58:30.706326454Z" level=error msg="Failed to delete a new service for vip 10.255.0.4 fwmark 2865: no such process"
Sep 27 20:58:30 test01 dockerd[99]: time="2016-09-27T20:58:30.713753137Z" level=info msg="setting up rule failed, [-t nat -D DOCKER-INGRESS -p tcp --dport 8080 -j DNAT --to-destination <nil>:8080]:  (iptables failed: iptables --wait -t nat -D DOCKER-INGRESS -p tcp --dport 8080 -j DNAT --to-destination <nil>:8080: iptables v1.4.21: Bad IP address \"<nil>\"\n\nTry `iptables -h' or 'iptables --help' for more information.\n (exit status 2))"
```
Only a single entry will be in DOCKER-INGRESS chain (i.e. no DNAT entry)
```
# iptables -t nat -L DOCKER-INGRESS -n -v
Chain DOCKER-INGRESS (2 references)
 pkts bytes target     prot opt in     out     source               destination         
    1    64 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0 
```
A curl to the local service will fail:
```
# curl http://127.0.0.1:8080
curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
```

**Comment by virtuman** [comment]
@slawekgh is it safe to run docker on a host that has firewall disabled? I never considered it running docker without firewall on the host machine, but is there no reason not to do that?

Thanks

**Comment by slawekgh** [comment]
1. Ubuntu 16.04,  all swarm features works properly 
2. I think disabling firewall is not good idea at all,  no matter what app is running there

**Comment by rmdashrfslash** [comment]
Everything seems to be working for me on centos  7 with firewall D disabled and iptables enabled.

**Comment by virtuman** [comment]
@slawekgh I have a deployment of 3 ubuntu 16.04 servers and experiencing the exact same issue:

cat /etc/*rel*

> DISTRIB_ID=Ubuntu
> DISTRIB_RELEASE=16.04
> DISTRIB_CODENAME=xenial
> DISTRIB_DESCRIPTION="Ubuntu 16.04.1 LTS"
> NAME="Ubuntu"
> VERSION="16.04.1 LTS (Xenial Xerus)"
> ID=ubuntu
> ID_LIKE=debian
> PRETTY_NAME="Ubuntu 16.04.1 LTS"
> VERSION_ID="16.04"
> HOME_URL="http://www.ubuntu.com/"
> SUPPORT_URL="http://help.ubuntu.com/"
> BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
> UBUNTU_CODENAME=xenial


**docker info**
> Containers: 60
>  Running: 16
>  Paused: 0
>  Stopped: 44
> Images: 26
> Server Version: 1.12.1
> Storage Driver: aufs
>  Root Dir: /var/lib/docker/aufs
>  Backing Filesystem: extfs
>  Dirs: 334
>  Dirperm1 Supported: true
> Logging Driver: json-file
> Cgroup Driver: cgroupfs
> Plugins:
>  Volume: local
>  Network: overlay bridge null host
> Swarm: active
>  NodeID: c7g9fg9qnqdkj60f79mfk7gd7
>  Is Manager: true
>  ClusterID: 9gs3d6e5djcrbq0ut6p706olh
>  Managers: 3
>  Nodes: 3
>  Orchestration:
>   Task History Retention Limit: 5
>  Raft:
>   Snapshot Interval: 10000
>   Heartbeat Tick: 1
>   Election Tick: 3
>  Dispatcher:
>   Heartbeat Period: 5 seconds
>  CA Configuration:
>   Expiry Duration: 3 months
>  Node Address: 138.201.122.195
> Runtimes: runc
> Default Runtime: runc
> Security Options: apparmor seccomp
> Kernel Version: 4.4.0-24-generic
> Operating System: Ubuntu 16.04.1 LTS
> OSType: linux
> Architecture: x86_64
> CPUs: 8
> Total Memory: 62.75 GiB
> Name: srv-eu2
> ID: 4LTP:DI4R:HUWB:LSRA:WSMQ:XBJD:YJJ3:7U5Q:FX7C:EKFT:WTYS:UCGV
> Docker Root Dir: /var/lib/docker
> Debug Mode (client): false
> Debug Mode (server): false
> Registry: https://index.docker.io/v1/
> WARNING: No swap limit support
> Insecure Registries:
>  127.0.0.0/8

**Comment by virtuman** [comment]
i take it back, on ubuntu i was able to get things to work after modifying:
you can either disable UFW OR change /etc/default/ufw to:
DEFAULT_FORWARD_POLICY="ACCEPT"

seems to work now on ubuntu, centos still no go :(

**Comment by sanimej** [comment]
@slawekgh 
`on Ubuntu 16.04 worked for me without touching firewalld and iptables `
By this do yo mean you had firewalld running and still the routing mesh was working correctly on all nodes ?

**Comment by sanimej** [comment]
@virtuman I am not sure why removing the iptables REJECT rules manually didn't help in your case. If disabling firewalld (but have iptables enabled) is an option you can do that. I will have to spend some time to figure out how to have firewalld allow 4789/7946 ports. Just doing `--add-port` didn't help.

**Comment by slawekgh** [comment]
3 x fresh ubuntu 16.04 install , everything works fine, I created service with 2 containers, www served on port 9009 , on all 3 nodes I can curl 9009 tcp port and I get load-balancing working fine 

From default (after fresh install): 
ubuntu3@192.168.56.109# systemctl status firewalld
● firewalld.service
   Loaded: not-found (Reason: No such file or directory)
   Active: inactive (dead)

**Comment by virtuman** [comment]
@slawekgh it says here that you are checking status of firewalld but ubuntu 16.04 comes with ufw as the default, no ? 
firewalld is an option, but if ufw is stopped as well then you wouldn't have any issues, but if you start up ufw sevice, then things won't work until default forward policy is modified. Could you double check that in your test setup ? Thank you again 

```
root@srveu1 ~ # service ufw status
● ufw.service - Uncomplicated firewall
   Loaded: loaded (/lib/systemd/system/ufw.service; enabled; vendor preset: enabled)
   Active: active (exited) since Fri 2016-09-30 04:46:07 UTC; 18h ago
  Process: 28016 ExecStop=/lib/ufw/ufw-init stop (code=exited, status=0/SUCCESS)
  Process: 28183 ExecStart=/lib/ufw/ufw-init start quiet (code=exited, status=0/SUCCESS)
 Main PID: 28183 (code=exited, status=0/SUCCESS)
    Tasks: 0
   Memory: 0B
      CPU: 0
   CGroup: /system.slice/ufw.service

Sep 30 04:46:07 srveu1 systemd[1]: Starting Uncomplicated firewall...
Sep 30 04:46:07 srveu1 systemd[1]: Started Uncomplicated firewall.
```


@sanimej - I posted examples that I used for allowing port, but that rule wouldn't have affected my setup, since all of my nodes have 1 ethernet adaptor that is used for all "internal" networking and all traffic is completely open between nodes, so with that rule or without it - nothing is blocked and i can test it with:
UDP ports:
```
nc -vnzu 13.13.13.13 7946
nc -vnzu 13.13.13.13 4789
```
and TCP ports
`telnet 13.13.13.13 7946`

and servers in fact can communicate with each other perfectly fine, i'm guessing the culprit is something else, but that's just an FYI

**Comment by Sebastien-Meiffren** [comment]
Hi, 

I think I'm having the same issue on Ubuntu 16.04. I've installed a fresh server (dedicated server) yesterday at home with Ubuntu 16.04 & Docker 1.12.1 because I wanted to learn more about swarm !
I haven't done anything special on the configuration => nothing on IPTABLES, no FW.

````
chuivert@server1:~$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 22
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay null bridge host
Swarm: active
 NodeID: 0rwangehlqog5qew1u8clbb35
 Is Manager: true
 ClusterID: 77u5jevcyjmq9uu1bzaroz7dn
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.1.21
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.953 GiB
Name: server1
ID: V5HD:VVWF:VUKS:SODK:NTDG:5QSY:RFT6:7OD4:N7VK:3IEU:B7TW:O37X
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8

chuivert@server1:~$ sudo ufw status
Status: inactive

chuivert@server1:~$ docker service create   --publish 8080:80   --name nginx   nginx:latest

chuivert@server1:~$ docker service ls
ID            NAME   REPLICAS  IMAGE         COMMAND
4521zej1zo6j  nginx  1/1       nginx:latest

chuivert@server1:~$ docker service ps nginx
ID                         NAME     IMAGE         NODE     DESIRED STATE  CURRENT STATE          ERROR
3rr3o9qojdes9owzab0uvdp17  nginx.1  nginx:latest  server1  Running        Running 7 minutes ago
````

When I curled the url localhost:8080 a few seconds after creating the service. I got connection refused.
Now that it has been 7 min, I get stalled .... locally and remotely (I waited 3 min, no timeout). So I might have an issue but not related to this one, I don't know ! :/

**Comment by virtuman** [comment]
just upgraded to docker 1.12.2-rc3 and service appears to publish all ports correctly, if docker is started after firewalld service. Ie. if i do 
service firewalld restart
service docker restart

then things work correctly, although same notices appear under service firewalld status

but if i `service firewalld restart` after docker container was running - then all containers launched via docker service will become unavailable from this node again

**Comment by thaJeztah** [comment]
> but if i service firewalld restart after docker container was running - then all containers launched via docker service will become unavailable from this node again

That's a known issue; firewalld ignores rules that are set by docker, so if you restart firewalld, it will wipe all rules that docker sets.

Looks like this issue is resolved in 1.12.2 then?

**Comment by virtuman** [comment]
Actually it didn't allow me to successfully establish the connection to the service whether i started firewalld before or after up until 1.12.2-*

I wouldn't call the issue resolved, since successful operation depends on order of startup of services, and these errors still come up in the logs:

```
service firewalld status -l
Redirecting to /bin/systemctl status  -l firewalld.service
● firewalld.service - firewalld - dynamic firewall daemon
   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2016-10-10 01:35:11 UTC; 1 weeks 0 days ago
 Main PID: 25992 (firewalld)
   Memory: 20.2M
   CGroup: /system.slice/firewalld.service
           └─25992 /usr/bin/python -Es /usr/sbin/firewalld --nofork --nopid

Oct 17 06:55:35 www5.servers.com firewalld[25992]: 2016-10-17 06:55:35 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C DOCKER-ISOLATION -i docker_gwbridge -o docker0 -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Oct 17 06:55:36 www5.servers.com firewalld[25992]: 2016-10-17 06:55:36 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t mangle -C OUTPUT -p udp --dport 4789 -m u32 --u32 0>>22&0x3C@12&0xFFFFFF00=65536 -j MARK --set-mark 13681891' failed: iptables: No chain/target/match by that name.
Oct 17 06:55:36 www5.servers.com firewalld[25992]: 2016-10-17 06:55:36 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t mangle -C OUTPUT -p udp --dport 4789 -m u32 --u32 0>>22&0x3C@12&0xFFFFFF00=65792 -j MARK --set-mark 13681891' failed: iptables: No chain/target/match by that name.
Oct 17 06:55:41 www5.servers.com firewalld[25992]: 2016-10-17 06:55:41 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C DOCKER-INGRESS -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Oct 17 06:55:41 www5.servers.com firewalld[25992]: 2016-10-17 06:55:41 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t filter -C DOCKER-INGRESS -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Oct 17 06:55:41 www5.servers.com firewalld[25992]: 2016-10-17 06:55:41 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C POSTROUTING -m ipvs --ipvs -d 10.255.0.0/16 -j SNAT --to-source 10.255.0.3' failed: iptables: No chain/target/match by that name.
Oct 17 06:55:42 www5.servers.com firewalld[25992]: 2016-10-17 06:55:42 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C POSTROUTING -m ipvs --ipvs -d 10.0.0.0/24 -j SNAT --to-source 10.0.0.7' failed: iptables: No chain/target/match by that name.
Oct 17 06:56:17 www5.servers.com firewalld[25992]: 2016-10-17 06:56:17 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C POSTROUTING -m ipvs --ipvs -d 10.0.0.0/24 -j SNAT --to-source 10.0.0.7' failed: iptables: No chain/target/match by that name.
Oct 17 06:56:17 www5.servers.com firewalld[25992]: 2016-10-17 06:56:17 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C POSTROUTING -m ipvs --ipvs -d 10.0.0.0/24 -j SNAT --to-source 10.0.0.7' failed: iptables: No chain/target/match by that name.
Oct 17 06:56:45 www5.servers.com firewalld[25992]: 2016-10-17 06:56:45 ERROR: COMMAND_FAILED: '/sbin/iptables -w2 -t nat -C POSTROUTING -m ipvs --ipvs -d 10.0.0.0/24 -j SNAT --to-source 10.0.0.7' failed: iptables: No chain/target/match by that name.
```

**Comment by thaJeztah** [comment]
We made a change for the 1.13 release to set service order; https://github.com/docker/docker/pull/25592, but I'm not sure if there's more we _can_ do. Perhaps @sanimej has suggestions

**Comment by virtuman** [comment]
Everything stopped working altogether again :)

I also already tried adding same After condition in systemd unit, but to no avail.

Am I understanding this correctly - that ubuntu 16.04 is currently the best distribution as an OS to use on the host system to ensure that docker has most functionalities working properly out of the box? or should we be using something else? ie debian, coreos, rancheros ?

thank you.

**Comment by thaJeztah** [comment]
Ubuntu 16.04 is a good choice as it has a fairly recent kernel (which allows you, for example, to use the `overlay2` graphdriver), although there's many people running on CentOS/RHEL so I'm still wondering what causes this in your situation.

**Comment by virtuman** [comment]
i would definitely prefer to stick with centos7.2, i trust it much more then ubuntu. personal experience with ubuntu is very "sad" when comparing it against something like centos, which is absolutely not an issue when using on a host to run fully dockerized environment ...


Apparently my current issue is that containers don't see each other, ie:

```
docker network create -d overlay collabnet
docker service create --name wordpressdb --network collabnet -e MYSQL_ROOT_PASSWORD=mysql123 -e MYSQL_DATABASE=wordpress mysql:latest
docker service create -e WORDPRESS_DB_HOST=wordpressdb -e WORDPRESS_DB_PASSWORD=mysql123 --network collabnet --replicas 3 --name wordpressapp --publish 85:80 wordpress:latest
```

results in this:
```
docker exec -it wordpressapp.2.ejl1errhvnmthviwasw85xdku ping wordpressdb
PING wordpressdb (10.0.0.2): 56 data bytes
92 bytes from ba159a3f8b6b (10.0.0.7): Destination Host Unreachable
92 bytes from ba159a3f8b6b (10.0.0.7): Destination Host Unreachable
92 bytes from ba159a3f8b6b (10.0.0.7): Destination Host Unreachable
92 bytes from ba159a3f8b6b (10.0.0.7): Destination Host Unreachable
```

but checking the IP of container doesn't reveal any possible problem with access to it:

```
docker exec -it wordpressdb.1.4udfr70a425wbq6vah0i3l8f2 ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
883: eth0@if884: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default
    link/ether 02:42:0a:00:00:03 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.3/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet 10.0.0.2/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:aff:fe00:3/64 scope link
       valid_lft forever preferred_lft forever
885: eth1@if886: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:ac:12:00:03 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.3/16 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe12:3/64 scope link
       valid_lft forever preferred_lft forever
```

**Comment by sanimej** [comment]
@virtuman If you have firewalld disabled can you confirm if everything is working ok with 1.12.2 ? If that is the case we can sure the issue is only because of iptables conflict between firewalld and docker and look into that specifically.

**Comment by slawekgh** [comment]
LB is not for ICMP , only for TCP/UDP ?  :-)

**Comment by outofcoffee** [comment]
@virtuman this worked for us on RHEL 7.2 with 3.10 kernel:

```
firewall-cmd --permanent --zone=trusted --add-interface=docker_gwbridge
```

Published ports accessible outside of localhost.

YMMV.

**Comment by virtuman** [comment]
It does look like I can get to the web container, but wordpress cannot resolve wordpressdb 

ie:

```
MySQL Connection Error: (2002) Connection timed out
Warning: mysqli::mysqli(): (HY000/2002): Connection timed out in - on line 19
```

**Comment by virtuman** [comment]
Upgraded to docker 1.12.3, still having same DNS issues where wordpress container cannot connect to mysql container when both launched using commands described in this ticket on centos 7.2 hosts  (6 hosts in the swarm now) :

```
docker network create -d overlay collabnet
docker service create --name wordpressdb --network collabnet -e MYSQL_ROOT_PASSWORD=mysql123 -e MYSQL_DATABASE=wordpress mysql:latest
docker service create -e WORDPRESS_DB_HOST=wordpressdb -e WORDPRESS_DB_PASSWORD=mysql123 --network collabnet --replicas 3 --name wordpressapp --publish 85:80/tcp wordpress:latest
```

any thoughts on how else i can try to troubleshoot this further?

**Comment by sanimej** [comment]
@virtuman Can you confirm that you have the connectivity issues only when firewalld is enabled ?

**Comment by virtuman** [comment]
yes, confirmed. 
with firewalld disabled everything seems to work ok


Comment from @outofcoffee was spot on. it didnt' do the trick for me but it got me looking in the direction of firewalld zones and what i found was the following:

1. needed to assign proper interfaces to appropriate zones:  
  1. trusted (active)  interfaces: docker0 docker_gwbridge eth1 
  2. public (active) interfaces: eth0
2. had to look into --permanent vs non-permanent diffs, found that for some reason had to configure all interfaces for proper zones,although I assumed that --permanent is the one that sticks after the reboot, but in my case, for some reason non-permanent was reloaded alongside with mismatched --permanent zone rules.

now things seem to work, but i'm afraid to switch production to start using docker services since firewalld is not something that i feel comfortable with debugging.

I guess the issue is somehow related to the zones, would be great if docker could sample all needed zone rulesets, and at this point I'm not really clear on how it would need to be setup if the server had only 1 network interface in it .. super confused, off to read firewalld manuals

**Comment by virtuman** [comment]
will there be any deeper automation for firewalld rules / zones management, to make sure that docker works properly out of the box, or is this not something on your backlog for a while ?

**Comment by mauricioszabo** [comment]
This is happening with me on ARM. I'm using docker on CHIP (the $9 computer) and it doesn't publish my service at all. I'm using the following code:

`$ docker service create --replicas 3 --name pokemon --publish 8080:3000 mauricioszabo/pokemon`

but it doesn't publish any port. But when I run it with:

`$ docker run -p8081:3000 -it mauricioszabo/pokemon`

It published port 8081 correctly.

The `mauricioszabo/pokemon` image is a multi-arch test image that runs on x86_84 and ARM. The worse, and most strange thing, is that when I create a swarm with ARM and x86 nodes, the ports are only published on x86 nodes. This gives the **very strange** behavior that, sometimes, when I probe for port 8080, it reports open and other times, it reports closed (because it's routing to ARM machines).

---

My system data:

chip@chip:~$ uname -a
Linux chip 4.4.13-ntc-mlc #1 SMP Tue Dec 6 21:38:00 UTC 2016 armv7l GNU/Linux
chip@chip:~$ docker -v
Docker version 17.03.0-ce, build 60ccb22

**Comment by thaJeztah** [comment]
@mauricioszabo when publishing a service's ports, the published ports are not visible in `docker ps`, because the containers backing the service are not published directly, but addressed through the load-balancer. If you `docker service inspect <service-name>`, you should see the port-mapping though

**Comment by mauricioszabo** [comment]
@thaJeztah I know, and that's not my problem. The problem is that, for `docker service create` on ARM machines, it is not publishing the port. With docker inspect, I see the port mapping but still the problem persists.

I tried some debugging tools and found out that I'm missing a kernel module, so I'll try again.

I just found strange that, when I'm using swarm mode with two machines, one that's publishing the port correctly and other that it is not, the manager is still trying to route to the machine that can't publish the port.

**Comment by mauricioszabo** [comment]
Same problem with a ARM linux with every required kernel option activated...

**Comment by DBLaci** [comment]
I had a similar problem on Gentoo. I enabled IP_VS_RR in the kernel, and the swarm overlay network is not working fine.
This thread lead me to the solution:

https://github.com/scaleway/kernel-tools/issues/293
